From d286a8d005dfa997d6e2e80997460006a43769b3 Mon Sep 17 00:00:00 2001
From: Ken <21211439+kencu@users.noreply.github.com>
Date: Sat, 20 Aug 2022 16:49:47 -0400
Subject: [PATCH 1/4] SnowLeopard Fixes

---
 dlls/winex11.drv/pen.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/pen.c b/dlls/winex11.drv/pen.c
index b1ad1026056..de7340b550d 100644
--- a/dlls/winex11.drv/pen.c
+++ b/dlls/winex11.drv/pen.c
@@ -47,7 +47,7 @@ static DWORD get_user_dashes( char *res, const DWORD *style, DWORD len )
         }
         else dashes[pos++] = dashes[i];
     }
-    for (i = 0; i < pos; i++) res[i] = min( dashes[i], 255 );
+    for (i = 0; i < pos; i++) res[i] = min( dashes[i], (int)255 );
     return pos;
 }
 
-- 
2.37.1 (Apple Git-137.1)

