From 00b5080760faee4142a959f3dd23f322b0c9377b Mon Sep 17 00:00:00 2001
From: Dean M Greer <38226388+Gcenx@users.noreply.github.com>
Date: Fri, 9 Jun 2023 15:26:00 -0400
Subject: [PATCH] loader: In macOS preloader, delay loading dylibs

---
 loader/preloader_mac.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/loader/preloader_mac.c b/loader/preloader_mac.c
index ae83dc9e..75940f4e 100644
--- a/loader/preloader_mac.c
+++ b/loader/preloader_mac.c
@@ -299,6 +299,7 @@ void *wld_munmap( void *start, size_t len );
 SYSCALL_FUNC( wld_munmap, 73 /* SYS_munmap */ );
 
 static intptr_t (*p_dyld_get_image_slide)( const struct target_mach_header* mh );
+static void (*p_dyld_make_delayed_module_initializer_calls)(void);
 
 #define MAKE_FUNCPTR(f) static typeof(f) * p##f
 MAKE_FUNCPTR(dlopen);
@@ -635,6 +636,7 @@ void *wld_start( void *stack, int *is_unix_thread )
     LOAD_POSIX_DYLD_FUNC( dlopen );
     LOAD_POSIX_DYLD_FUNC( dlsym );
     LOAD_POSIX_DYLD_FUNC( dladdr );
+    LOAD_MACHO_DYLD_FUNC( _dyld_make_delayed_module_initializer_calls );
     LOAD_MACHO_DYLD_FUNC( _dyld_get_image_slide );
 
 #ifdef __i386__ /* CrossOver Hack #16371 */
@@ -664,6 +666,8 @@ void *wld_start( void *stack, int *is_unix_thread )
     if (!map_region( &builtin_dlls ))
         builtin_dlls.size = 0;
 
+    p_dyld_make_delayed_module_initializer_calls();
+
     /* load the main binary */
     if (!(mod = pdlopen( argv[1], RTLD_NOW )))
         fatal_error( "%s: could not load binary\n", argv[1] );
