diff --git a/dlls/win32u/vulkan.c b/dlls/win32u/vulkan.c
index 37b37e84d8a..de54ac8e035 100644
--- a/dlls/win32u/vulkan.c
+++ b/dlls/win32u/vulkan.c
@@ -3028,14 +3028,21 @@ static void vulkan_init_once(void)
     VkExtensionProperties *properties = NULL;
     uint32_t count = 0;
     VkResult res;
+    const char *libvulkan, *graphics_backend;
 
-#ifdef SONAME_LIBVULKAN
-    vulkan_handle = dlopen( SONAME_LIBVULKAN, RTLD_NOW );
-    if (!vulkan_handle) ERR( "Failed to load %s\n", SONAME_LIBVULKAN );
-#else
-    ERR( "Wine was built without Vulkan support.\n" );
-#endif
-    if (!vulkan_handle) return;
+    /* CW HACK 25909: Allow specifying libvulkan */
+    if (!(libvulkan = getenv( "CX_LIBVULKAN" )) ||
+        !(graphics_backend = getenv( "CX_ACTIVE_GRAPHICS_BACKEND" )) ||
+        strcmp( graphics_backend, "wined3d" ))
+        libvulkan = SONAME_LIBVULKAN;
+    else
+        ERR( "Using %s\n", libvulkan );
+
+    if (!(vulkan_handle = dlopen( libvulkan, RTLD_NOW )))
+    {
+        ERR( "Failed to load %s: %s\n", libvulkan, dlerror() );
+        return;
+    }
 
 #define LOAD_FUNCPTR( f )                                                                          \
     if (!(p_##f = dlsym( vulkan_handle, #f )))                                                     \
