--- dlls/ntdll/signal_x86_64.c.orig.c	2019-09-13 16:32:49.000000000 -0400
+++ dlls/ntdll/signal_x86_64.c	2019-09-22 18:37:12.000000000 -0400
@@ -55,6 +55,9 @@
 # define UNW_LOCAL_ONLY
 # include <libunwind.h>
 #endif
+#ifdef __APPLE__
+# include <mach/mach.h>
+#endif
 
 #define NONAMELESSUNION
 #define NONAMELESSSTRUCT
@@ -3209,10 +3212,10 @@
     static int gsbase_offset = -1;
     void *ret;
 
-    kern_return_t kr = thread_info(mach_thread_self(), THREAD_IDENTIFIER_INFO, (thread_identifier_info_t) &tiinfo, &info_count);
+    kern_return_t kr = thread_info(mach_thread_self(), THREAD_IDENTIFIER_INFO, (thread_info_t) &tiinfo, &info_count);
     if (kr == KERN_SUCCESS)
     {
-        TRACE("pthread_self() %p thread ID %lx gsbase %lx\n", pthread_self(), tiinfo.thread_id, tiinfo.thread_handle);
+        TRACE("pthread_self() %p thread ID %llx gsbase %llx\n", pthread_self(), tiinfo.thread_id, tiinfo.thread_handle);
         return (void*)tiinfo.thread_handle;
     }
 
