From 461fa1b0bdd3bcdbed7191d86ab9fce14aba7b10 Mon Sep 17 00:00:00 2001
From: Gcenx <38226388+Gcenx@users.noreply.github.com>
Date: Fri, 22 Apr 2022 09:00:57 -0400
Subject: [PATCH] crypt32: expand cert search paths

---
 dlls/crypt32/unixlib.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/crypt32/unixlib.c b/dlls/crypt32/unixlib.c
index a345b5e6ac8..67be04892ad 100644
--- a/dlls/crypt32/unixlib.c
+++ b/dlls/crypt32/unixlib.c
@@ -612,6 +612,7 @@ static void import_certs_from_path(LPCSTR path, BOOL allow_dir)
 }
 
 static const char * const CRYPT_knownLocations[] = {
+ "@@PREFIX@@/share/curl/curl-ca-bundle.crt", /* Macports */
  "/etc/ssl/certs/ca-certificates.crt",
  "/etc/ssl/certs",
  "/etc/pki/tls/certs/ca-bundle.crt",
--
2.37.1 (Apple Git-137.1)
