--- dlls/winegstreamer/mf_decode.c.orig.c    2020-08-15 14:07:11.000000000 -0400
+++ dlls/winegstreamer/mf_decode.c    2020-08-15 14:08:33.000000000 -0400
@@ -1048,7 +1048,7 @@
 
     drain = gst_query_new_drain();
     gst_pad_peer_query(decoder->input_src, drain);
-    gst_query_unref (query);
+    gst_query_unref(drain);
 
     if (decoder->output_counter || decoder->draining)
     {
@@ -1353,4 +1353,4 @@
             return;
         }
     }
-}
\ No newline at end of file
+}
