From 7eddd79e4e9831e42bc30a8ace55d9e533510684 Mon Sep 17 00:00:00 2001
From: Brendan Shanks <bshanks@codeweavers.com>
Date: Tue, 8 Sep 2020 16:09:46 -0700
Subject: [PATCH] winemac.drv: Print error when no GPUs are detected.

Signed-off-by: Brendan Shanks <bshanks@codeweavers.com>
Signed-off-by: Ken Thomases <ken@codeweavers.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/winemac.drv/display.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git dlls/winemac.drv/display.c dlls/winemac.drv/display.c
index a96a7b2c5f6..078cb55ab8a 100644
--- dlls/winemac.drv/display.c
+++ dlls/winemac.drv/display.c
@@ -1790,6 +1790,8 @@ void macdrv_init_display_devices(BOOL force)
     if (macdrv_get_gpus(&gpus, &gpu_count))
         goto done;
     TRACE("GPU count: %d\n", gpu_count);
+    if (!gpu_count)
+        ERR("No GPUs detected\n");
 
     for (gpu = 0; gpu < gpu_count; gpu++)
     {
