--- ./dlls/gdi32/freetype.c.orig.c	2019-08-23 10:29:32.000000000 -0400
+++ ./dlls/gdi32/freetype.c	        2019-09-29 16:09:21.000000000 -0400
@@ -169,6 +169,7 @@
 #ifdef FT_LCD_FILTER_H
 static FT_Error (*pFT_Library_SetLcdFilter)(FT_Library, FT_LcdFilter);
 #endif
+static FT_Error (*pFT_Property_Set)(FT_Library, const FT_String *, const FT_String *, const void *);
 
 #ifdef SONAME_LIBFONTCONFIG
 #include <fontconfig/fontconfig.h>
@@ -4161,6 +4162,8 @@
 #ifdef FT_LCD_FILTER_H
     pFT_Library_SetLcdFilter = wine_dlsym(ft_handle, "FT_Library_SetLcdFilter", NULL, 0);
 #endif
+    pFT_Property_Set = wine_dlsym(ft_handle, "FT_Property_Set", NULL, 0);
+
 
     if(pFT_Init_FreeType(&library) != 0) {
         ERR("Can't init FreeType library\n");
@@ -4174,6 +4177,9 @@
     FT_SimpleVersion = ((FT_Version.major << 16) & 0xff0000) |
                        ((FT_Version.minor <<  8) & 0x00ff00) |
                        ((FT_Version.patch      ) & 0x0000ff);
+    
+    FT_UInt interpreter_version = 35;
+    pFT_Property_Set( library, "truetype", "interpreter-version", &interpreter_version );
 
     font_driver = &freetype_funcs;
     return TRUE;
