From b99ec04f557fad83b89be406cbc605f09fdcc18a Mon Sep 17 00:00:00 2001
From: Gijs Vermeulen <gijsvrm@gmail.com>
Date: Fri, 7 Aug 2020 16:20:23 +0200
Subject: [PATCH] configure: Add an option to disable QuickTime support.

Signed-off-by: Gijs Vermeulen <gijsvrm@gmail.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 configure    | 44 +++++++++++++++++++++++++++++++-------------
 configure.ac | 37 +++++++++++++++++++++----------------
 2 files changed, 52 insertions(+), 29 deletions(-)

diff --git configure configure
index c93e000e8f9..f2785a280f5 100755
--- configure
+++ configure
@@ -878,6 +878,7 @@ with_pcap
 with_png
 with_pthread
 with_pulse
+with_quicktime
 with_sane
 with_sdl
 with_tiff
@@ -2608,6 +2609,7 @@ Optional Packages:
   --without-png           do not use PNG
   --without-pthread       do not use the pthread library
   --without-pulse         do not use PulseAudio sound support
+  --without-quicktime     do not use QuickTime support
   --without-sane          do not use SANE (scanner support)
   --without-sdl           do not use SDL
   --without-tiff          do not use TIFF
@@ -4114,6 +4116,12 @@ if test "${with_pulse+set}" = set; then :
 fi
 
 
+# Check whether --with-quicktime was given.
+if test "${with_quicktime+set}" = set; then :
+  withval=$with_quicktime;
+fi
+
+
 # Check whether --with-sane was given.
 if test "${with_sane+set}" = set; then :
   withval=$with_sane;
@@ -8977,13 +8985,16 @@ done
 
         LIBS="$ac_save_LIBS"
     fi
-    if test "$ac_cv_header_QuickTime_ImageCompression_h" = "yes"
+
+        if test "x$with_quicktime" != "xno"
     then
-                ac_save_LIBS="$LIBS"
-        LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo $LIBS"
-        { $as_echo "$as_me:${as_lineno-$LINENO}: checking for the QuickTime framework" >&5
+        if test "$ac_cv_header_QuickTime_ImageCompression_h" = "yes"
+        then
+                        ac_save_LIBS="$LIBS"
+            LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo $LIBS"
+            { $as_echo "$as_me:${as_lineno-$LINENO}: checking for the QuickTime framework" >&5
 $as_echo_n "checking for the QuickTime framework... " >&6; }
-        cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+            cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 #include <QuickTime/ImageCompression.h>
 int
@@ -8997,22 +9008,29 @@ _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
   { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
-                        QUICKTIME_LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo"
+                            QUICKTIME_LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo"
 
-                        enable_wineqtdecoder=${enable_wineqtdecoder:-yes}
+                            enable_wineqtdecoder=${enable_wineqtdecoder:-yes}
 else
   { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
 $as_echo "no" >&6; }
-                        ac_cv_header_QuickTime_ImageCompression_h=no
+                            ac_cv_header_QuickTime_ImageCompression_h=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
-        LIBS="$ac_save_LIBS"
-    fi
-    if test "$ac_cv_header_QuickTime_ImageCompression_h" != "yes"
-    then
-        as_fn_append wine_notices "|QuickTime ${notice_platform}development files not found, video decoding won't be supported."
+            LIBS="$ac_save_LIBS"
+        fi
     fi
+    if test "$ac_cv_header_QuickTime_ImageCompression_h" != "yes"; then :
+  case "x$with_quicktime" in
+  x)   as_fn_append wine_notices "|QuickTime ${notice_platform}development files not found, video decoding won't be supported." ;;
+  xno) ;;
+  *)   as_fn_error $? "QuickTime ${notice_platform}development files not found, video decoding won't be supported.
+This is an error since --with-quicktime was requested." "$LINENO" 5 ;;
+esac
+
+fi
+
     if test "$ac_cv_header_Carbon_Carbon_h" = "yes"
     then
         CARBON_LIBS="-framework Carbon"
diff --git configure.ac configure.ac
index 4ff0f9c81f7..0f4b0b817a9 100644
--- configure.ac
+++ configure.ac
@@ -79,6 +79,7 @@ AC_ARG_WITH(png,       AS_HELP_STRING([--without-png],[do not use PNG]))
 AC_ARG_WITH(pthread,   AS_HELP_STRING([--without-pthread],[do not use the pthread library]),
             [if test "x$withval" = "xno"; then ac_cv_header_pthread_h=no; fi])
 AC_ARG_WITH(pulse,     AS_HELP_STRING([--without-pulse],[do not use PulseAudio sound support]))
+AC_ARG_WITH(quicktime, AS_HELP_STRING([--without-quicktime],[do not use QuickTime support]))
 AC_ARG_WITH(sane,      AS_HELP_STRING([--without-sane],[do not use SANE (scanner support)]))
 AC_ARG_WITH(sdl,       AS_HELP_STRING([--without-sdl],[do not use SDL]))
 AC_ARG_WITH(tiff,      AS_HELP_STRING([--without-tiff],[do not use TIFF]))
@@ -830,24 +831,28 @@ case $host_os in
         AC_CHECK_FUNCS(IOHIDManagerCreate)
         LIBS="$ac_save_LIBS"
     fi
-    if test "$ac_cv_header_QuickTime_ImageCompression_h" = "yes"
-    then
-        dnl Make sure we can actually use the QuickTime framework
-        ac_save_LIBS="$LIBS"
-        LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo $LIBS"
-        AC_MSG_CHECKING([for the QuickTime framework])
-        AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <QuickTime/ImageCompression.h>]],[[CodecManagerVersion(NULL);]])],
-                       [AC_MSG_RESULT([yes])
-                        AC_SUBST(QUICKTIME_LIBS,"-framework QuickTime -framework ApplicationServices -framework CoreVideo")
-                        enable_wineqtdecoder=${enable_wineqtdecoder:-yes}],
-                       [AC_MSG_RESULT([no])
-                        ac_cv_header_QuickTime_ImageCompression_h=no])
-        LIBS="$ac_save_LIBS"
-    fi
-    if test "$ac_cv_header_QuickTime_ImageCompression_h" != "yes"
+
+    dnl *** Check for QuickTime ***
+    if test "x$with_quicktime" != "xno"
     then
-        WINE_NOTICE([QuickTime ${notice_platform}development files not found, video decoding won't be supported.])
+        if test "$ac_cv_header_QuickTime_ImageCompression_h" = "yes"
+        then
+            dnl Make sure we can actually use the QuickTime framework
+            ac_save_LIBS="$LIBS"
+            LIBS="-framework QuickTime -framework ApplicationServices -framework CoreVideo $LIBS"
+            AC_MSG_CHECKING([for the QuickTime framework])
+            AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <QuickTime/ImageCompression.h>]],[[CodecManagerVersion(NULL);]])],
+                           [AC_MSG_RESULT([yes])
+                            AC_SUBST(QUICKTIME_LIBS,"-framework QuickTime -framework ApplicationServices -framework CoreVideo")
+                            enable_wineqtdecoder=${enable_wineqtdecoder:-yes}],
+                           [AC_MSG_RESULT([no])
+                            ac_cv_header_QuickTime_ImageCompression_h=no])
+            LIBS="$ac_save_LIBS"
+        fi
     fi
+    WINE_NOTICE_WITH(quicktime,[test "$ac_cv_header_QuickTime_ImageCompression_h" != "yes"],
+                     [QuickTime ${notice_platform}development files not found, video decoding won't be supported.])
+
     if test "$ac_cv_header_Carbon_Carbon_h" = "yes"
     then
         AC_SUBST(CARBON_LIBS,"-framework Carbon")
